[Merged by Bors] - feat: persist nipost state efficiently

0xjac opened the pull request:
## Motivation
<!-- Please mention the issue fixed by this PR or detailed motivation -->
Closes #4506
<!-- `Closes #XXXX, closes #XXXX, ...` links mentioned issues to this PR and automatically closes them when this it's merged -->

## Changes
<!-- Please describe in detail the changes made -->
- write: compute CRC as the file is written. (requires two writes: content & CRC)
- read: compute CRC as the file is read.
- use CRC Size constant from CRC pkg.
- *extra: Properly close the files when reading and writing*

The changes match what was requested in #4506 with regards to storing NiPoST state.

## Test Plan
<!-- Please specify how these changes were tested 
(e.g. unit tests, manual testing, etc.) -->

Unit tests `TestWriteCRC` & `TestReadCRC` were added to specifically test the writing and reading of valid and invalid CRCs.  
Note: The write tests do not rely on `read()` and the read tests do not rely on `write()`.

I am unsure if a more elaborate plan is needed.

## Documentation
As far as I can see, there is no documentation related to the NiPoST state storage format.

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [ ] This PR does not make changes to log messages (which monitoring infrastructure may rely on)
  This PR changes error messages when writing and reading NiPoST states from file. Specifically:
    - write: new message `create temporary file <path>: <err>`
    - write: replace `save file <path>: err` with `save file from <tmpfile>, <path>: <err>`
    - read: replace `read file <path>: <err>` with `open file <path>: <err>`
    - read: new message `failed to get file info <path>: <err>`
    - read: change in `wrong checksum <read_cksum>, computed <computed_cksum>`: checksums are displayed in (uppercase) hex rather than decimal.

0xjac committed:
feat: persist nipost state efficiently

- write: compute CRC as the file is written. (requires two writes:
  content & CRC)
- read: compute CRC as the file is read.
- use CRC Size constant from CRC pkg.

Closes #4506

0xjac commented:
Regarding the change of format for the files. If keeping the content of the existing files is requested, I think the best solution is to have some kind of migration logic where the files are read the old way and persisted the new way.

Also note: the issue #4506 talks about NiPoST states but also mentions other places where data is persisted to disk. I had a brief look and found that:
- `p2p/book/book.go` has some [persistence logic using CRC-64](https://github.com/spacemeshos/go-spacemesh/blob/develop/p2p/book/book.go#L389-L439)
- `p2p/identity.go` has a [TODO to add a CRC-64 check](https://github.com/spacemeshos/go-spacemesh/blob/develop/p2p/identity.go#L52)

However those seem unrelated to the NiPoST state. Let me know if you want me to implement CRC properly for those too in a separaate PR.

dshulyak commented on activation/nipost_state.go:
```diff
@@ -20,43 +19,64 @@ const (
 	binary.BigEndian.PutUint64(crc, checksum.Sum64())
-	if _, err := buf.Write(crc); err != nil {
-		return fmt.Errorf("write checksum %v: %w", path, err)
+	if _, err = tmp.WriteAt(crc, 0); err != nil {
```
is it a common practice to add crc at the start? what are the disadvantages of adding it as a suffix? i think crc as a suffix allows to preserve sequential writes, it obviously doesn't matter here

0xjac replied:
The [first point of the specification on Wikipedia](https://en.wikipedia.org/wiki/Cyclic_redundancy_check) seems to indicate it is. To be taken with a grain of salt.

That being said I would argue having some fixed length data first and the variable length data afterwards makes for a better design. especially if the variable length data is really large. It is also the way it was specified in the issue.

Also looking further into it, adding a zeros prefix does not change the CRC so I can rewrite the `write` function to remove the initial write of the prefix.

Or I can rewrite it with the CRC as a suffix. The logic might be a bit more complicated for the read because we need the logic to stop before the end of the file but that can work too.

dshulyak reviewed (changes requested):
> Regarding the change of format for the files. If keeping the content of the existing files is requested, I think the best solution is to have some kind of migration logic where the files are read the old way and persisted the new way.

i think it would be needed, it is a pain to update othrwise. but also in order to simplify deployment i would run migration on go-spacemesh startup. e.g it can look up files with version 1 and migrate them to version 2. otherwise good change

we are merging last minute cleanups and preparations for release, once it is out, this change will be merged as well.

0xjac commented:
> > Regarding the change of format for the files. If keeping the content of the existing files is requested, I think the best solution is to have some kind of migration logic where the files are read the old way and persisted the new way.
> 
> i think it would be needed, it is a pain to update othrwise. but also in order to simplify deployment i would run migration on go-spacemesh startup. e.g it can look up files with version 1 and migrate them to version 2. otherwise good change
> 
> we are merging last minute cleanups and preparations for release, once it is out, this change will be merged as well.

The hard part is those files have no concept of versioning so I can't just run something at startup every time. The migration has to happen during the update of the software (which might be platform dependent).

I'm not sure exactly how you proceed in those cases. I can think of the following solutions:
- You have some mechanism to run code when upgrading `go-spacemesh`
- You have some mechanism to run an update/migration only once on startup and not every time.
- Create a separate command and document that the user run it after updating the software.

Let me know which way you want to go about this?

Edit: this is only relevant if CRC is a prefix. As pointed in the original issue, it is a pain to change the format after the launch. But if time is short and you don't want to break anything you could keep it as a suffix and once you have some sort of migration mechanism in place between versions of spacemesh itself, then you can move the CRC as a prefix.

dshulyak commented:
> The hard part is those files have no concept of versioning so I can't just run something at startup every time. The migration has to happen during the update of the software (which might be platform dependent).

i think the most future proof approach is to add a version. to run the first migration we can have some heuristic, like if after crc we have `{` then it is a version 0. sometimes migrations won't be possible at all, as content of the file is used as a challenge for downstream protocols, and we will have to keep several versions in the codebase/state

regarding upgrades i think we want to keep it simple for now and try to do all of it on spacemesh startup.

@fasmat what do you think?

0xjac commented:
Regarding version information, I would include that as part of the "content" so the version information is also taken into account for the CRC. So that would move it out of scope for this issue. The location of the CRC (rpefix/suffix) would be a lower level issue than file version.

Another quick fix, tho I'm not sure you would like it is to change the filenames (and you could even include version info in there). That would allow to read the files properly based on the name. Something like:
- `nipost_challenge.bin` → `nipost_challenge.2.bin`
- `nipost_builder_state.bin` → `nipost_builder_state.2.bin`
- `post.bin` → `post.2.bin`

And if the names bother you, in the future once everyone is using the `.2.bin` you can rename it back to `.bin`.

fasmat commented:
Thanks for your contribution @0xjac!

@dshulyak  The main challenge that I see here is that `write` and `read` are used for multiple data types and files. In the long run we need a better way to generalize our persistence of data. I think `write` and `read` should take `io.Writer`/`io.Reader` and `scale.Encodable`/`scale.Decodable` instead of a file name and `[]byte`.

This also has the advantage that the caller of those functions can read/write a few bytes before passing the arguments to set or get the version or other metadata of the file or data type. This would also make changes to the file format in the future easier.

If we want to include this change we should already add a version to the new file, to avoid migrating a file twice and to improve from there. Yes, your suggestion to do it at start up keeps the scope of this change limited if we decide to include it before genesis.

@0xjac I think adding a few bytes at the beginning of the file (for now maybe just encoding 0x0001 as first version) is better than renaming the files, otherwise we might need to keep track of 2 filenames per file in the future.

0xjac commented:
Happy to add a version at the beginning if this is what you want but:
- `0x0001` do you want a whole 4 bytes for the version (i.e. over 4 billion possible versions)?
- How would you approach the issues of current files which do not have a version prefix? Especially on the unlikely event that the start of the content matches a known version prefix?

Overall, I would consider versioning for the content and the encoding method of the data. But CRC happens below that, once your versioned content is encoded into bytes and needs to be saved to a file. The way it should work, I think, is:
1. Split the CRC and content.
2. Compute the content's CRC
3. Compare the computed CRC with the one of the file
4. Pass the content up and only at this point look for versioning and the decoded content, etc.

Steps 1 to 3 should not really change unless the way integrity is handled changes (very unlikely). Version changes should happen if you switch from something like json to cbor or change the structure of the data.

I don't know how much of a pain rebuilding those files are. But my take would be screw it. Do CRC properly before launch as a prefix (and maybe sneak in the version there too) and if some users have to recompute some data even if it is a pain it's only for testnet and on mainnet at least it will be fine. And we can provide a small script (happy to do it) for testnet users to manually run this time to update their file so they don't have to recompute anything.

@fasmat regarding your comment on a better way to generalize our persistence of data in the long run. I agree. Happy to give it a shot after the merge if you want. I would also suggest to add a mechanism to perform some changes on updates similar to db migration but more generic as on the long run I am sure some things will change and this will be needed. (At this point, handling breaking changes of the persistence logic should be easier.)

0xjac commented:
@dshulyak @fasmat 

> Also looking further into it, adding a zeros prefix does not change the CRC so I can rewrite the write function to remove the initial write of the prefix.

I was wrong here. So the only advantage of having the checksum first is that the logic to read it is a bit easier. Hence, screw it. Let's keep it at the end and keep compatibility with previous format.

I've updated the code to have the checksum at the end. The only extra logic and cost is the `fstat` syscall to get the file size. This is needed to stop reading as we reach the checksum.

Since the format is the same, there is no need for versioning or anything at this point. This can be tackled later, esp. as genesis is near.

This PR only improves the efficiency of the read/write with respect to the checksum computation. Let me know what you think.

*edit 1: Updated PR description to reflect changes.*
*edit 2: The existing code was not explicitly closing the file after reading/writing. This is now fixed too.*

fasmat commented:
Thanks for all your effort @0xjac 🙂 

The code looks good to me and keeping compatibility with the existing implementation, should make integration easier. However, since we are so close to Genesis and this code is used in multiple places, we won't be able to test the change in time for it to be included in the release.

However I would like to merge it after Genesis and include it in the next release 🙂

fasmat commented:
bors try

fasmat commented:
@0xjac besides the linter complaining the CI looks good. Could you run `make lint-fix` on your branch and commit the changes? 🙂

0xjac commented:
@fasmat Done. Sorry about that my IDE took my global linter not the local one and it missed a couple things.

Note; I've also rebased the branch on `upstream/develop`

fasmat commented:
bors merge

dshulyak commented on activation/nipost_state.go:
```diff
@@ -20,43 +19,71 @@ const (
+	}
+
+	saved := make([]byte, crc64.Size)
+	if _, err = file.Read(saved); err != nil {
```
won't it fail with io unexpected eof here for existing files without crc? can we please add such test?

fasmat replied:
The code stats the file, and first reads all data except the last 8 (`crc64.size`) bytes from the file into data. Then it reads the remaining 8 bytes into `saved` and compares if saved matches the calculated CRC.

There can only be an `io.EOF` here if the file changes size while the node is reading it 😕

0xjac replied:
Yes, if the CRC is missing the last 8 bits will be read as CRC and then the checksum verification will fail. I'll add a quick test for that case.

If you want a test where the file is changed between the reads I can try to make one but that might be a bit trickier. Because we pass the path as a string. But if you want I can mock the filesystem for `read` and test it.

dshulyak reviewed (approved):
thanks, ignore my last question i confused things

fasmat commented:
bors merge

fasmat commented:
bors merge

fasmat commented:
The change seems to fail because the `/tmp` directory on the nodes we use for system tests is on a different device than the path the file ultimately is written to. `atomic.ReplaceFile` fails with the following error: "rename /tmp/nipost_challenge.bin283715628 /data/post/nipost_challenge.bin: invalid cross-device link"

I assume we have to create the file we write to in the same directory as path for this to work as we expect it. Probably replacing `os.CreateTemp("", filepath.Base(path))` with `os.Create(fmt.Sprintf("%s.tmp", path))` should fix the issue.

dshulyak commented:
btw do we store this nipost state next to post data? if so it is reasonable to assume that tmp will be on different disk, so it is not only systest specific

0xjac commented:
I'm away from my laptop at the moment but I can fix it maybe tonight or Sunday

fasmat commented:
> btw do we store this nipost state next to post data? if so it is reasonable to assume that tmp will be on different disk, so it is not only systest specific

yes and agree 🙂

fasmat committed:
Create temporary file in same directory

fasmat commented:
bors try

fasmat committed:
Merge remote-tracking branch 'origin/develop' into fix/4506-persist-nipost-state

fasmat commented:
bors try

fasmat commented:
bors merge
